pytorch_lightning/trainer/trainer.py

```diff
@@ -1231,38 +1221,40 @@ def _call_teardown_hook(self) -> None:
         # summarize profile results
         self.profiler.describe()
 
-    def call_hook(self, hook_name: str, *args, **kwargs) -> Any:
-        if self.lightning_module:
-            prev_fx_name = self.lightning_module._current_fx_name
-            self.lightning_module._current_fx_name = hook_name
+    def call_hook(
+        self, hook_name: str, *args: Any, pl_module: Optional["pl.LightningModule"] = None, **kwargs: Any
+    ) -> Any:
+        pl_module = self.lightning_module or pl_module
+        if pl_module:
+            prev_fx_name = pl_module._current_fx_name
+            pl_module._current_fx_name = hook_name
 
         # always profile hooks
         with self.profiler.profile(hook_name):
 
             # first call trainer hook
-            if hasattr(self, hook_name):
-                trainer_hook = getattr(self, hook_name)
-                trainer_hook(*args, **kwargs)
+            callback_fx = getattr(self, hook_name, None)
+            if callable(callback_fx):
+                callback_fx(*args, **kwargs)
 
             # next call hook in lightningModule
             output = None
-            model_ref = self.lightning_module
-            if is_overridden(hook_name, model_ref):
-                hook_fx = getattr(model_ref, hook_name)
-                output = hook_fx(*args, **kwargs)
+            model_fx = getattr(pl_module, hook_name, None)
+            if callable(model_fx):
+                output = model_fx(*args, **kwargs)
```

Comment:
what are the implications? is it safe to call the default implementation? there must have been a reason in the past why we had is_overridden.

Reply:
Keeping the `is_overridden` check, so it doesn't call the base implementation

Reply:
Actually, just remembered why I need this.

The following tests fail with the `is_overridden` check:

```python
FAILED tests/callbacks/test_finetuning_callback.py::test_finetuning_callback
FAILED tests/callbacks/test_stochastic_weight_avg.py::test_swa_warns - ValueE...
FAILED tests/trainer/test_dataloaders.py::test_request_dataloader - Attribute...
FAILED tests/trainer/test_states.py::test_trainer_fn_while_running[Single-Step]
FAILED tests/trainer/test_trainer.py::test_model_in_correct_mode_during_stages[None-1]
FAILED tests/trainer/test_trainer.py::test_model_in_correct_mode_during_stages[ddp_cpu-2]
```

This is because the `on_*_model_{train,eval}` hooks do provide a default implementation so we need to call them even if they are not overridden.

Related to https://github.com/PyTorchLightning/pytorch-lightning/discussions/8760